### PR TITLE
The Doc update for ScheduleDaemonSetPods

### PR DESCRIPTION
**Pull request by k82cn**
Signed-off-by: Da K. Ma <klaus1982.cn@gmail.com>

### CODE REVIEW

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 1]
```diff
@@ -117,18 +152,22 @@ they will not be evicted when there are node problems such as a network partitio
 `TaintBasedEvictions` feature is not enabled, they are also not evicted in these scenarios, but
 due to hard-coded behavior of the NodeController rather than due to tolerations).
 
- They also tolerate following `NoSchedule` taints:
+They also tolerate following `NoSchedule` taints:
 
- - `node.kubernetes.io/memory-pressure`
  - `node.kubernetes.io/disk-pressure`
+ - `node.kubernetes.io/memory-pressure`
+ - `node.kubernetes.io/unschedulable` (after version 1.10)
```
Comment (k82cn) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: The PR is merged in 1.10, although the doc is update in 1.11 :(

**Comment by mistyhacks** [comment]
This is not a 1.11 feature, so please just open this PR against `master`. I'm removing the `1.11` milestone.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 2]
```diff
@@ -93,6 +93,7 @@ different Kubernetes components.
 | `TokenRequest` | `false` | Alpha | 1.10 | |
 | `VolumeScheduling` | `false` | Alpha | 1.9 | 1.9 |
 | `VolumeScheduling` | `true` | Beta | 1.10 | |
+| `ScheduleDaemonSetPods` | `false` | Alpha | 1.11 | |
```
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md]: - version incorrect
- mind the alpha-beta order
- provide a short description for the gate

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 3]
```diff
@@ -117,18 +152,22 @@ they will not be evicted when there are node problems such as a network partitio
...
 When the support to critical pods is enabled and the pods in a DaemonSet are
 labeled as critical, the Daemon pods are created with an additional
 `NoSchedule` toleration for the `node.kubernetes.io/out-of-disk` taint.
 
-Note that all above `NoSchedule` taints above are created only in version 1.8 or later if the alpha feature `TaintNodesByCondition` is enabled.
+Note that all above `NoSchedule` taints are created only in version 1.8 or later
+ if the alpha feature `TaintNodesByCondition` is enabled.
+
+Also note that the `node-role.kubernetes.io/master` `NoSchedule` toleration specified in the above example
+ is needed in version 1.6 or later to schedule on *master* nodes, because this is not a default toleration.
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: are you sure we want to mention things that old?

**Comment by mistyhacks** [comment]
I rebased to resolve conflicts.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 4]
```diff
@@ -106,6 +108,39 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
  - The DaemonSet controller can make Pods even when the scheduler has not been started, which can help cluster
    bootstrap.
 
+### Scheduled by default scheduler (with ScheduleDaemonSetPods alpha feature)
+
+In version 1.11, `ScheduleDaemonSetPods` was introduced as an alpha feature. With this feature,
+DaemonSet Pods are scheduled by default scheduler, instead of Daemonset controller.
```
Comment (bsalamat) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: s/Daemonset/DaemonSet/
to be consistent with the rest of the doc

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 5]
```diff
@@ -106,6 +108,39 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+        - target-host-name
+```
+
+When this feature is enabled, `node.kubernetes.io/unschedulable:NoSchedule` toleration is added automatically to
+DaemonSet Pods. DaemonSet controller was ignoring `unschedulable` Nodes when scheduling DaemonSet Pods.
+In order to ensure that default scheduler keeps the same behavior and schedules DaemonSet Pods on `unschedulable` nodes,
+`TaintNodesByCondition` must be enabled.
+
+When this feature and `TaintNodesByCondition` are enabled together, `node.kubernetes.io/network-unavailable:NoSchedule`
+toleration is required to DaemonSet pods that using host network.
```
Comment (bsalamat) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: This sentence is unclear. Could you please rephrase?

Comment (k82cn) [reply content/en/docs/concepts/workloads/controllers/daemonset.md]: Something like this: When this feature and `TaintNodesByCondition` are enabled together, if DaemonSet uses host network, `node.kubernetes.io/network-unavailable:NoSchedule` toleration is required. ?

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 6]
```diff
@@ -94,6 +94,7 @@ different Kubernetes components.
...
```
Comment (janetkuo) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md]: There are two `ScheduleDaemonSetPods` in this table. Did you revert the doc change for this feature in 1.10?

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 7]
```diff
@@ -117,18 +153,22 @@ they will not be evicted when there are node problems such as a network partitio
...
+ - `node.kubernetes.io/memory-pressure`
+ - `node.kubernetes.io/unschedulable` (after version 1.10)
 
 When the support to critical pods is enabled and the pods in a DaemonSet are
 labeled as critical, the Daemon pods are created with an additional
 `NoSchedule` toleration for the `node.kubernetes.io/out-of-disk` taint.
 
-Note that all above `NoSchedule` taints above are created only in version 1.8 or later if the alpha feature `TaintNodesByCondition` is enabled.
+Note that all above `NoSchedule` taints are created only in version 1.8 or later
+ if the alpha feature `TaintNodesByCondition` is enabled.
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: This "Taints and Tolerations" section is becoming difficult to follow. 

Would you make it easier to read, such as using tables and clarify what is required when which feature is enabled?

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 8]
```diff
@@ -106,6 +108,40 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+
+When this feature is enabled, `node.kubernetes.io/unschedulable:NoSchedule` toleration is added automatically to
+DaemonSet Pods. DaemonSet controller was ignoring `unschedulable` Nodes when scheduling DaemonSet Pods.
+In order to ensure that default scheduler keeps the same behavior and schedules DaemonSet Pods on `unschedulable` nodes,
+`TaintNodesByCondition` must be enabled.
+
+When this feature and `TaintNodesByCondition` are enabled together, `node.kubernetes.io/network-unavailable:NoSchedule`
+toleration is required to DaemonSet pods that using host network.
+
+
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: Worth adding the reason why we need this feature / benefit of using default scheduler.

File: content/en/docs/concepts/configuration/taint-and-toleration.md [hunk 9]
```diff
@@ -210,6 +210,7 @@ currently include:
  * `node.kubernetes.io/memory-pressure`: Node has memory pressure.
```
Comment (janetkuo) [inline content/en/docs/concepts/configuration/taint-and-toleration.md]: In L200: "Kubernetes 1.6 has alpha support for representing node problems."

Is it still alpha support?

File: content/en/docs/concepts/configuration/taint-and-toleration.md [hunk 10]
```diff
@@ -210,6 +210,7 @@ currently include:
...
  * `node.kubernetes.io/disk-pressure`: Node has disk pressure.
  * `node.kubernetes.io/network-unavailable`: Node's network is unavailable.
+ * `node.kubernetes.io/unschedulable`: Node is unschedulable. (1.10 or later)
  * `node.cloudprovider.kubernetes.io/uninitialized`: When kubelet is started
    with "external" cloud provider, it sets this taint on a node to mark it
    as unusable. When a controller from the cloud-controller-manager initializes
```
Comment (janetkuo) [inline content/en/docs/concepts/configuration/taint-and-toleration.md]: Is `TaintBasedEvictions` still alpha?

Comment (k82cn) [reply content/en/docs/concepts/configuration/taint-and-toleration.md]: Yes, `TaintBasedEvictions` is still alpha right now .

**Comment by mistyhacks** [comment]
Please address all the technical review comments and then let me know when this is ready to be copyedited. Thanks!

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 11]
```diff
@@ -106,6 +108,40 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+  requiredDuringSchedulingIgnoredDuringExecution:
+    nodeSelectorTerms:
+    - matchFields:
+      - key: metadata.name
+        operator: In
+        values:
+        - target-host-name
+```
+
+When this feature is enabled, `node.kubernetes.io/unschedulable:NoSchedule` toleration is added automatically to
```
[the remark this answers is not in the document]
Comment (k82cn) [reply content/en/docs/concepts/workloads/controllers/daemonset.md]: In `ScheduleDaemonSetPods`, if want to schedule a DaemonSet pod to unschedulable nodes, we has to let default scheduler can tolerate it. As `TaintNodesByCondition` is alpha, it must be enabled; but I'd like to upgrade it ASAP :)

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 12]
```diff
@@ -106,6 +108,40 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+DaemonSet Pods are scheduled by default scheduler, instead of DaemonSet controller.
+
+When this feature is enabled, DaemonSet controller creates all DaemonSet pods upfront no matter whether they can be
+scheduled right away or not; and the `NodeAffinity` term (instead of `.spec.nodeName`) is added to those DaemonSet Pods,
+as in the following example. This enables the default scheduler to bind the Pod to the target host. If node affinity of
+DaemonSet Pod already exists, it will be replaced. DaemonSet controller will only perform these operations when creating
+DaemonSet Pods; and those operations will only modify the Pods of DaemonSet, no changes are made to the `.spec.template`
+of DaemonSet.
+
+```
```
[the remark this answers is not in the document]
Comment (k82cn) [reply content/en/docs/concepts/workloads/controllers/daemonset.md]: Done

File: content/en/docs/concepts/workloads/controllers/daemonset.md (hunk 5 above)
Comment (k82cn) [reply content/en/docs/concepts/workloads/controllers/daemonset.md]: Yes, that's true. DaemonSet controller should add this toleration automatically to keep behaviour consistent. I'll open an PR for this :)

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 6 above)
Comment (k82cn) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md]: Removed; not sure how this happened. The old PR was closed instead of merged.

File: content/en/docs/concepts/workloads/controllers/daemonset.md (hunk 8 above)
Comment (k82cn) [reply content/en/docs/concepts/workloads/controllers/daemonset.md]: Done

**Comment by k82cn** [comment]
@janetkuo , PTAL :)

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 13]
```diff
@@ -106,29 +108,64 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+| `node.kubernetes.io/not-ready`           | NoExecute  | `TaintBasedEvictions`                                        | 1.8+    | when `TaintBasedEvictions`  is enabled,they will not be evicted when there are node problems such as a network partition. |
+| `node.kubernetes.io/unreachable`         | NoExecute  | `TaintBasedEvictions`                                        | 1.8+    | when `TaintBasedEvictions`  is enabled,they will not be evicted when there are node problems such as a network partition. |
+| `node.kubernetes.io/disk-pressure`       | NoSchedule | `TaintNodesByCondition`                                      | 1.8+    |                                                              |
+| `node.kubernetes.io/memory-pressure`     | NoSchedule | `TaintNodesByCondition`                                      | 1.8+    |                                                              |
+| `node.kubernetes.io/unschedulable`       | NoSchedule | `ScheduleDaemonSetPods`, `TaintNodesByCondition`             | 1.11+   | When ` ScheduleDaemonSetPods` is enabled, ` TaintNodesByCondition` is necessary to make sure DaemonSet pods tolerate unscheuable attributes by default scheduler. |
```
Comment (tobru) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: `s/unscheuable/unschedulable/`

**Comment by mistyhacks** [comment]
@k82cn ping, this still has quite a few unaddressed comments and now has a conflict that needs to be resolved. PTAL.

**Comment by k82cn** [comment]
done, let me ping janetku@ for tech review :)

File: content/en/docs/concepts/configuration/taint-and-toleration.md (hunk 9 above)
Comment (janetkuo) [reply content/en/docs/concepts/configuration/taint-and-toleration.md]: It seems that we don't have alpha support taints anymore. 
"... Kubernetes 1.6 has alpha support for representing node problems" --> "... Kubernetes taints nodes that have problems."

File: content/en/docs/concepts/configuration/taint-and-toleration.md [hunk 14]
```diff
@@ -204,12 +204,13 @@ currently include:
...
  * `node.kubernetes.io/not-ready`: Node is not ready. This corresponds to
    the NodeCondition `Ready` being "`False`".
- * `node.alpha.kubernetes.io/unreachable`: Node is unreachable from the node
+ * `node.kubernetes.io/unreachable`: Node is unreachable from the node
    controller. This corresponds to the NodeCondition `Ready` being "`Unknown`".
  * `node.kubernetes.io/out-of-disk`: Node becomes out of disk.
  * `node.kubernetes.io/memory-pressure`: Node has memory pressure.
  * `node.kubernetes.io/disk-pressure`: Node has disk pressure.
  * `node.kubernetes.io/network-unavailable`: Node's network is unavailable.
+ * `node.kubernetes.io/unschedulable`: Node is unschedulable. (1.10 or later)
```
Comment (janetkuo) [inline content/en/docs/concepts/configuration/taint-and-toleration.md]: Question for docs review: do we still need `(1.10 or later)` given that docs are now versioned?

File: content/en/docs/concepts/configuration/taint-and-toleration.md (hunk 9 above)
Comment (janetkuo) [reply content/en/docs/concepts/configuration/taint-and-toleration.md]: "Taint based Evictions (alpha feature)" in L186 needs to be updated as well.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 15]
```diff
@@ -106,29 +108,64 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
  - The DaemonSet controller can make Pods even when the scheduler has not been started, which can help cluster
    bootstrap.
 
-Daemon Pods do respect [taints and tolerations](/docs/concepts/configuration/taint-and-toleration),
-but they are created with `NoExecute` tolerations for the following taints with no `tolerationSeconds`:
+### Scheduled by default scheduler (with ScheduleDaemonSetPods alpha feature)
+
+A DaemonSet ensures that all (or some) nodes run a copy of a pod. Normally, the machine that a pod runs on is selected by the Kubernetes scheduler; however, pods of DaemonSet are created and scheduled by DaemonSet controller who leveraged kube-scheduler’s predicates policy. That introduces the following issues:
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: Remove "who leveraged kube-scheduler’s predicates policy"; this is implementation details.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 16]
```diff
@@ -106,29 +108,64 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+
+ * DaemonSet can not respect Node’s resource changes, e.g. more resources after other Pods exit ([#46935](https://github.com/kubernetes/kubernetes/issues/46935), [#58868](https://github.com/kubernetes/kubernetes/issues/58868))
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: Remove this. It's more of a DS controller bug.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 17]
```diff
@@ -106,29 +108,64 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+ * DaemonSet can not respect Pod Affinity and Pod AntiAffinity ([#29276](https://github.com/kubernetes/kubernetes/issues/29276))
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: Remove this. This can be fixed by adding predicates to DS controller.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 18]
```diff
@@ -106,29 +108,64 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+ * Duplicated logic to respect scheduler features, e.g. critical pods ([#42028](https://github.com/kubernetes/kubernetes/issues/42028)),  tolerant/taint
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: Remove this; users don't care about duplicated logic in our code base.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 19]
```diff
@@ -106,29 +108,64 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+ * Hard to debug why DaemonSet’s Pod is not created, e.g. not enough resources; it’s better to have a pending Pods with predicates’ event
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: Rephrase:
* Inconsistent Pod behavior: normal Pods waiting to be scheduled will be created and in `Pending` state, but DaemonSet pods in this case won't be created. It can cause user confusions. 

DaemonSet controller will still emit for unscheduleable pods, so it's still discoverable by users.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 20]
```diff
@@ -106,29 +108,64 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+ * Hard to support preemption in different components, e.g. DS and default scheduler
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: Would you rephrase it to target users instead of kube-developers?

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 21]
```diff
@@ -106,29 +108,64 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+
+In version 1.11, `ScheduleDaemonSetPods` was introduced as an [alpha feature](/docs/reference/command-line-tools-reference/feature-gates/). With this feature,
+DaemonSet Pods are scheduled by default scheduler, instead of DaemonSet controller.
+
+When this feature is enabled, DaemonSet controller creates all DaemonSet pods upfront no matter whether they can be
+scheduled right away or not; and the `NodeAffinity` term (instead of `.spec.nodeName`) is added to those DaemonSet Pods,
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: "no matter whether they can be scheduled right away or not" is ambiguous. 
Need to clarify that DaemonSet controller still respects scheduling policy specified in the spec.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 22]
```diff
@@ -106,29 +108,64 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+| `node.kubernetes.io/memory-pressure`     | NoSchedule | `TaintNodesByCondition`                                      | 1.8+    |                                                              |
+| `node.kubernetes.io/unschedulable`       | NoSchedule | `ScheduleDaemonSetPods`, `TaintNodesByCondition`             | 1.11+   | When ` ScheduleDaemonSetPods` is enabled, ` TaintNodesByCondition` is necessary to make sure DaemonSet pods tolerate unschedulable attributes by default scheduler. |
+| `node.kubernetes.io/network-unavailable` | NoSchedule | `ScheduleDaemonSetPods`, `TaintNodesByCondition`, hostnework | 1.11+   | When ` ScheduleDaemonSetPods` is enabled, ` TaintNodesByCondition` is necessary to make sure DaemonSet pods, who uses host network, tolerate network-unavailable attributes by default scheduler. |
+| `node.kubernetes.io/out-of-disk`         | NoSchedule | `ExperimentalCriticalPodAnnotation` (critical pod only), `TaintNodesByCondition` | 1.8+    |                                                              |
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: This is much better. Thanks!

File: content/en/docs/concepts/configuration/taint-and-toleration.md (hunk 9 above)
Comment (k82cn) [reply content/en/docs/concepts/configuration/taint-and-toleration.md]: `TaintBasedEviction` is still alpha :( , I'm trying to graduate it to beta in one or two release.

**Comment by mistyhacks** [comment]
I added a copyedit. PTAL @k82cn

**Comment by k82cn** [comment]
@mistyhacks, thanks, t he copyedit LGTM :),

**Comment by zparnold** [comment]
Is this ready to go now?

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 23]
```diff
@@ -106,29 +108,74 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+
+{{< feature-state state="alpha" for-kubernetes-version="1.11"}}
+
+A DaemonSet ensures that all eligible nodes run a copy of a Pod. Normally, the
+node that a Pod runs on is selected by the Kubernetes scheduler. However,
+DaemonSet pods are created and scheduled by the DaemonSet controller instead.
+That introduces the following issues:
+
+ * A DaemonSet cannot respect a Node’s resource changes, such as the node
+   having more free resources after other Pods exit ([#46935](https://github.com/kubernetes/kubernetes/issues/46935)
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: Remove this as this isn't directly related.

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 24]
```diff
@@ -106,29 +108,74 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+ * Inconsistent Pod behavior: Normal Pods waiting to be scheduled are created
+   and in `Pending` state, but DaemonSet pods are not created in `Pending`
+   state. This is confusing to the user.
+ * With preemption, different components, such as DaemonSets and the default
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: It'd be helpful to link to the preemption doc, such as this one: https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/

File: content/en/docs/concepts/workloads/controllers/daemonset.md [hunk 25]
```diff
@@ -106,29 +108,74 @@ when the Pod is created, so it is ignored by the scheduler).  Therefore:
...
+   scheduler, may make different preemption decisions, which can cause
+   unexpected behaviorss.
```
Comment (janetkuo) [inline content/en/docs/concepts/workloads/controllers/daemonset.md]: Rephrase it to:

* Pod preemption is handled by default scheduler. When preemption is enabled, the DaemonSet controller will make scheduling decisions without considering pod priority and preemption.

**Comment by k82cn** [comment]
Done, @mistyhacks, please help to review the latest commit :)

**Comment by janetkuo** [comment]
Tech review is done. You may want to squash commits.

**Comment by mistyhacks** [comment]
I squashed and fixed a missing `>` in a feature-state shortcode. Let's see if Netlify doesn't fail anymore.

**Comment by k82cn** [comment]
> I squashed and fixed a missing

@mistyhacks , thanks very much for your help, you're so nice :)
